File: python/gridsearch.py
import numpy as np
import pandas as pd

# Import dataset
df = pdf.read_csv('data.csv')
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

# Split into training and test sets
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Examples for different models
method = 'knn'
if method == 'knn':
	from sklearn.neighbors import KNeighborsClassifier
	model = KNeighborsClassifier(n_neighbors=k)

	# Test every odd k from 1 to 51
	grid = [{'n_neighbors': list(range(1, 51, 2))}]

else if method == 'svmRadial':
	from sklearn.svm import SVR
	model = SVR(kernel='rbf')

	# Standard deviation, soft penalty, margin for SVM
	grid = [{'kernel': ['rbf'],
					 'gamma': [1e-4, 1e-3, 0.01, 0.1, 1, 10, 100],
					 'C': [1, 10, 100, 1000],
					 'epsilon': [0, 0.01, 0.1, 0.2, 0.5, 1, 2, 4]}]

# Apply grid search to model method with parameters in grid
from sklearn.grid_search import GridSearchCV
grid_search = GridSearchCV(model, param_grid=grid, cv=10)
grid_search.fit(X_train, y_train)

# Report best
print(grid_search.best_params_)

# Report score
print("Cross-validation score %0.2f" % grid_search.best_score_)

File: r/gridsearch.r
# Import dataset
df = read.csv('data.csv')

# caTools contains several basic utility functions including: moving 
# (rolling, running) window statistic functions, read/write for GIF 
# and ENVI binary files, fast calculation of AUC, LogitBoost classifier, 
# base64 encoder/decoder, round-off-error-free sum and cumsum, etc.
install.packages('caTools')
library(caTools)

# Create Boolean vector with random split
# SplitRatio specifies proportion of training set
split = sample.split(df$y, SplitRatio=0.8)

# Split into training and test set
train = subset(df, split == TRUE)
test = subset(df, split == FALSE)

# caret: Classification and Regression Training provides misc functions for 
# training and plotting classification and regression models.
install.packages('caret')
library(caret)

# Apply 10 folds cross-validation
folds = 10
tc = trainControl(method='cv', number=folds)

# Examples for different models
method = 'knn'
if (method == 'knn') {
	
	# Test every odd k from 1 to 51
	grid = expand.grid(k=seq.int(1, 51, 2))

} else if (method == 'svmRadial') {

	# Standard deviation, soft penalty, margin for SVM
	grid = expand.grid(sigma=c(1e4, 1e3, 100, 10, 1, 0.1, 0.01),
										 C=c(1, 10, 100, 1000),
										 epsilon=c(0, 0.01, 0.1, 0.2, 0.5, 1, 2, 4))
}

# Apply grid search to model method with parameters in grid
fit = train(y ~ ., data=train,
						method=method,
						control=tc,
						preProcess=c("center", "scale"),
						tuneGrid=grid

# Report scores
fit

# Plot scores
plot(fit)